### ISSUE TITLE
Unable to do a passwordless ssh to the remote esxi server

### ISSUE THREAD

**Issue by skorada-rvbd** [josenk/vagrant-vmware-esxi#12]
Hi,

I am trying to do a passwordless ssh to the esxi from my host machine but I am always prompted for the password even though it has been mentioned in the Vagrantfile. 

*plugin version*
vagrant-vmware-esxi (1.5.0)

Issue was also seen on 1.4.0

*Trail 1: with setting the password*

```
Vagrantfile

VAGRANTFILE_API_VERSION = "2"
Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
  config.ssh.forward_agent = true
  config.vm.define test-vm do |x|
    x.vm.box = "bento/ubuntu-16.04"
    x.vm.provider :vmware_esxi do |v|
      v.esxi_hostname = '10.73.40.19'
      v.esxi_username = 'root'
      v.esxi_password = 'secret'   =====> PASSWORD SET
      v.custom_vmx_settings = [['vhv.enable', 'TRUE']]
      v.vmname = test-vm
      v.memsize = 6000
      v.numvcpus = 2
    end
  end
end

Console Logs

username@hostname:~/home/vagrant$ sudo vagrant up
Bringing machine 'test_vm' up with 'vmware_esxi' provider...
==> test-vm: Virtual Machine will be built.
==> test-vm: --- WARNING         : config.virtual_network not set, using VM Network
==> test-vm: ---
==> test-vm: --- ESXi host       : 10.73.40.19
==> test-vm: --- VM Name         : test-vm
==> test-vm: --- Box             : bento/ubuntu-16.04
==> test-vm: --- Box Ver         : 201801.02.0
==> test-vm: --- Disk Store      : datastore1
==> test-vm: --- CPUS            : 2
==> test-vm: --- Memsize (MB)    : 6000
==> test-vm: --- Guest OS type   : ubuntu-64
==> test-vm: --- Virtual Network : ["VM Network"]
==>test-vm: --- Resource Pool   :
VMware ovftool 4.2.0 (build-5965791)
Opening VMX source: /home/username/.vagrant.d/boxes/bento-VAGRANTSLASH-ubuntu-16.04/201801.02.0/vmware_desktop/ZZZZ_test-vm.vmx
Enter login information for target vi://10.73.40.19/
Username: root
Password: **********  =====> PASSWORD PROMPT
Opening VI target: vi://root@10.73.40.19:443/
Deploying to VI: vi://root@10.73.40.19:443/
```

*Trial 2: with password in a file*
```
Vagrantfile:

VAGRANTFILE_API_VERSION = "2"
Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
  config.ssh.forward_agent = true
  config.vm.define test-vm do |x|
    x.vm.box = "bento/ubuntu-16.04"
    x.vm.provider :vmware_esxi do |v|
      v.esxi_hostname = '10.73.40.19'
      v.esxi_username = 'root'
      v.esxi_password = 'file:/tmp/esxi_password'   =====> PASSWORD SET THROUGH FILE
      v.custom_vmx_settings = [['vhv.enable', 'TRUE']]
      v.vmname = test-vm
      v.memsize = 6000
      v.numvcpus = 2
    end
  end
end

username@host-name:~/home/vagrant$ ls -l /tmp/esxi_password
-rw-r--r-- 1 root root 11 Jan 26 10:00 /tmp/esxi_password
username@host-name:~/home/vagrant$ cat /tmp/esxi_password
secret

Console logs:

username@host-name:~/home/vagrant$ sudo vagrant up
Bringing machine 'test-vm' up with 'vmware_esxi' provider...
==> test-vm: Virtual Machine will be built.
==> test-vm: --- WARNING         : config.virtual_network not set, using VM Network
==> test-vm: ---
==> test-vm: --- ESXi host       : 10.73.40.19
==> test-vm: --- VM Name         : test-vm
==> test-vm: --- Box             : bento/ubuntu-16.04
==> test-vm: --- Box Ver         : 201801.02.0
==> test-vm: --- Disk Store      : datastore1
==> test-vm: --- CPUS            : 2
==> test-vm: --- Memsize (MB)    : 6000
==> test-vm: --- Guest OS type   : ubuntu-64
==> test-vm: --- Virtual Network : ["VM Network"]
==> test-vm: --- Resource Pool   :
VMware ovftool 4.2.0 (build-5965791)
Opening VMX source: /home/username/.vagrant.d/boxes/bento-VAGRANTSLASH-ubuntu-16.04/201801.02.0/vmware_desktop/ZZZZ_test-vm.vmx
Enter login information for target vi://10.73.40.19/
Username: root
Password: **********   =====> PASSWORD PROMPT
Opening VI target: vi://root@10.73.40.19:443/
Deploying to VI: vi://root@10.73.40.19:443/

```

Am I missing something or using it the wrong way?
Only after entering the password the deployment starts. Is there way to do a passwordless ssh to the esxi server?

Regards

**Comment by josenk** [comment]
Yes, you are using the password configuration correctly.      Actually the paswordless ssh part IS working.  The console output shows that ovftool is asking for the password.    Most likely there's another special character I need to encode???   I never found a definitive list of characters I need to encode.   My suggestion is to add debugging.  **The following WILL expose your password to the console, so don't post the output here until you review it...**  You also might want to try changing your password to figure out which character is causing problems.    Add a debug line to the provider section of the Vagrantfile and try again.  

```
   v.debug = 'true password'
```

Let me know, Jonathan

**Comment by skorada-rvbd** [comment]
Hi Jonathan,

I have tried what you suggested and added v.debug = 'true' in the provider section

```

Bringing machine 'test-vm' up with 'vmware_esxi' provider...
==> test-vm: Virtual Machine will be built.
Available DS Volumes: ["datastore1"]
Available Networks: ["VM Network", "Management Network"]
==> test-vm: --- WARNING         : config.virtual_network not set, using VM Network
guestvm_network[0]: VM Network
==> test-vm: ---
==> test-vm: --- ESXi host       : 10.73.40.19
==> test-vm: --- VM Name         : test-vm
==> test-vm: --- Box             : bento/ubuntu-16.04
==> test-vm: --- Box Ver         : 201801.02.0
==> test-vm: --- Disk Store      : datastore1
==> test-vm: --- CPUS            : 2
==> test-vm: --- Memsize (MB)    : 6000
==> test-vm: --- Guest OS type   : ubuntu-64
==> test-vm: --- Virtual Network : ["VM Network"]
==> test-vm: --- Resource Pool   :
VMware ovftool 4.2.0 (build-5965791)
ovftool command: ovftool --noSSLVerify   --net:"net0=VM Network" -dm=thin  -ds="datastore1" --name="test-vm" "/home/username/.vagrant.d/boxes/bento-VAGRANTSLASH-ubuntu-16.04/201801.02.0/vmware_desktop/ZZZZ_test-vm.vmx" vi://root:******@10.73.40.19

Opening VMX source: /home/username/.vagrant.d/boxes/bento-VAGRANTSLASH-ubuntu-16.04/201801.02.0/vmware_desktop/ZZZZ_test-vm.vmx
Enter login information for target vi://10.73.40.19/
Username: root
Password: ====> OVFTOOL not reading the password?
```

As per the debug logs the ovftool command does show that the username and password for the server is being passed.Maybe due to the special character (I have a '@' in my password) it is having an issue.

Will also have try changing the password and update the thread.

Thanks
Shashank

**Comment by josenk** [comment]
Try 
```
v.debug = 'true password'
```

**Comment by skorada-rvbd** [comment]
Assuming my password is hello@123.
this gets passed on as hello%2540123
the '@' is encoded as %2540

```
Bringing machine 'test-vm' up with 'vmware_esxi' provider...
==> test-vm: Virtual Machine will be built.
Available DS Volumes: ["datastore1"]
Available Networks: ["VM Network", "Management Network"]
==> test-vm: --- WARNING         : config.virtual_network not set, using VM Network
guestvm_network[0]: VM Network
==> test-vm: ---
==> test-vm: --- ESXi host       : 10.73.40.19
==> test-vm: --- VM Name         : test-vm
==> test-vm: --- Box             : bento/ubuntu-16.04
==> test-vm: --- Box Ver         : 201801.02.0
==> test-vm: --- Disk Store      : datastore1
==> test-vm: --- CPUS            : 2
==> test-vm: --- Memsize (MB)    : 6000
==> test-vm: --- Guest OS type   : ubuntu-64
==> test-vm: --- Virtual Network : ["VM Network"]
==> test-vm: --- Resource Pool   :
VMware ovftool 4.2.0 (build-5965791)
ovftool command: ovftool --noSSLVerify   --net:"net0=VM Network" -dm=thin  -ds="datastore1" --name="test-vm" "/home/username/.vagrant.d/boxes/bento-VAGRANTSLASH-ubuntu-16.04/201801.02.0/vmware_desktop/ZZZZ_test-vm.vmx" vi://root:hello%2540123@10.73.40.19
ovftool command: ovftool --noSSLVerify   --net:"net0=VM Network" -dm=thin  -ds="datastore1" --name="test-vm" "/home/username/.vagrant.d/boxes/bento-VAGRANTSLASH-ubuntu-16.04/201801.02.0/vmware_desktop/ZZZZ_test-vm.vmx" vi://root:******@10.73.40.19
Opening VMX source: /home/username/.vagrant.d/boxes/bento-VAGRANTSLASH-ubuntu-16.04/201801.02.0/vmware_desktop/ZZZZ_test-vm.vmx
Enter login information for target vi://10.73.40.19/
Username: root
Password:
```

**Comment by josenk** [comment]
OK, I see what's happening...   The @ is getting encoded to %40, then the % from the encoding is getting encoded to %25.  It will be fixed in the next update.

**Comment by josenk** [comment]
Fixed in V1.5.1

**Comment by skorada-rvbd** [comment]
Thanks. This is now working
